controllers/gitrepository_controller.go

```diff
@@ -500,19 +500,22 @@ func (r *GitRepositoryReconciler) reconcileArtifact(ctx context.Context,
 
 	// Set the ArtifactInStorageCondition if there's no drift.
 	defer func() {
-		if obj.GetArtifact().HasRevision(artifact.Revision) && !includes.Diff(obj.Status.IncludedArtifacts) {
+		if obj.GetArtifact().HasRevision(artifact.Revision) && !includes.Diff(obj.Status.IncludedArtifacts) && !obj.IgnoreHasChanged() {
 			conditions.Delete(obj, sourcev1.ArtifactOutdatedCondition)
 			conditions.MarkTrue(obj, sourcev1.ArtifactInStorageCondition, meta.SucceededReason,
 				"stored artifact for revision '%s'", artifact.Revision)
 		}
 	}()
 
 	// The artifact is up-to-date
-	if obj.GetArtifact().HasRevision(artifact.Revision) && !includes.Diff(obj.Status.IncludedArtifacts) {
+	if obj.GetArtifact().HasRevision(artifact.Revision) && !includes.Diff(obj.Status.IncludedArtifacts) && !obj.IgnoreHasChanged() {
```

Comment:
It'd be good to have a test for it in https://github.com/fluxcd/source-controller/blob/39f1c5c217a55b07f053cc8d86e6eb8776da8b93/controllers/gitrepository_controller_test.go#L696 with some assertions in `afterFunc` to verify that the checksum is added in the status.

Same for the bucket controller change.

Reply:
I added two new test scenarios. Let me know if you see any other.